Vending step6

AppleCEO opened the pull request:
요구사항 반영하였습니다.

- 모델로부터 데이터를 받아서 뷰컨트롤러에서 뷰를 출력하던 것을 뷰컨트롤러에서 모델에 클로저를 보내서 처리
- 판매 목록을 이미지로 추가하도록 구현

AppleCEO committed:
수정 - 의미없는 공백 제거

AppleCEO committed:
수정 - 음료수 이미지 크기를 요구사항인 140x100 으로 변경

AppleCEO committed:
추가 - 이미지 라운드 추가

AppleCEO committed:
추가 - 구매 버튼 구성

AppleCEO committed:
제거 - Printable 관련 프로토콜은 한 파일에 통합

AppleCEO committed:
추가 - 판매 목록 출력을 위한 프로토콜 구현

AppleCEO committed:
추가 - 시작할 때 판매 목록을 출력하도록 함수 호출 추가

AppleCEO committed:
수정 - 잔고, 재고 목록 보이는 것도 클로저로 모델에 보내도록 수정

AppleCEO committed:
추가 - 13-6 결과 화면 캡처 추가

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -26,23 +33,40 @@ class ViewController: UIViewController {
+            { stock in
+                for (index, count) in stock.enumerated() {
+                    counts[index]?.text = String(count)+KoreanUnit.count.rawValue
+                }
```
프로토콜에서 함수를 추상화해서 동작을 만들때 매개변수가 배열로 그대로 넘기면 이렇게 내부 탐색을 직접 해야 하는 상황이 생깁니다. 
그래서 이렇게 배열 자체를 복사하도록 하기 보다는,
탐색을 할 수 있도록 배열 길이가 얼마인지, 배열 요소 중에서 필요한 값이 얼마인지 나눠서 호출하도록 세분화 하는게 좋습니다.
이렇게 배열을 가져오면 getter를 만드는 것과 다르지 않습니다. 
필요한 값을 가져오는게 아니라 좀더 구체적으로 어떤걸 화면에 출력할 것인지 그 동작을 추상화하는 게 좋습니다.

AppleCEO replied:
내부 탐색을 직접 하지 않도록 수정하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -26,23 +33,40 @@ class ViewController: UIViewController {
+    /// 이미지 Border, Round 변경
+    private func imageInit () {
+        for image in images {
+            image?.layer.cornerRadius = 15
```
이미지를 만들때도 ViewController가 이미지 속성을 바꿔주는 게 아니라
자기 속성을 바꾸는 RoundImageView 같은 구체적인 동작을 하는 뷰 객체를 만들고 담당 객체가 일을 하도록 개선하세요.

AppleCEO replied:
RoundImageView 객체를 만들고 음료 이미지들에 적용시켜봤습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -55,19 +79,71 @@ class ViewController: UIViewController {
+    @IBAction func buy(_ sender: UIButton) {
+        if sender.tag >= 0 && sender.tag <= 5 {
+            do {
+                try vendingMachine.buyToIndex(sender.tag)
```
이 부분도 구두로 피드백했던 내용입니다.
- sender.tag를 여기서 비교하기 보다는 메뉴를 처리하던 것처럼 enum 타입을 선언하던지, 선택하는 순서와 음료를 매칭하는 데이터 구조를 만들어서 기능을 담당하는 함수나 객체를 구현하는 게 좋습니다. 
그래서 vendingMachine에 넘길 때도 sender.tag를 넘기는 게 아니라 매칭된 음료를 넘기는게 낫지 않을까요? 
- 그리고 이런 구조에서 if 구문으로 들여쓰기를 하기 보다는 guard else가 더 적당한 것 같습니다. 
- 아래 에러가 나면 print()를 하고 있는데 이런 경우 앱에서는 어떻게 해야할까요? 사용자가 print() 결과를 보는게 아닙니다. 오류가 나면 사용자에게 어떻게 알려주는게 좋을지 생각해보세요.

AppleCEO replied:
- 순서와 음료를 매칭하는 enum 타입을 만들었습니다.
- 말씀하신 부분 guard else 로 변경하였습니다.
- UIAlertController 을 이용하여 오류창을 띄우도록 수정하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -55,19 +79,71 @@ class ViewController: UIViewController {
+                    let image1 = UIImage(named: imageName)
+                    let iv1 = UIImageView(image: image1!)
+                    iv1.frame = CGRect(x: xCoordinate, y: 575, width: 100, height: 100)
+                    self.view.addSubview(iv1)
```
구매 이력이 추가될 때마다 전체 목록을 다 가져와서 addSubview()를 하고 있습니다. 
이런 경우 이전에 추가했던 이미지 뷰는 어떻게 될까요? 
- 그러니까 [0], [0,1], [0,1,2] 이런 식으로 추가하다보면 `0` 이미지는 계속 누적되지 않을까요?
- `iv1` 같은 변수이름은 구체적이지 않습니다. 줄여서 쓰지말고 풀어서 써보세요. 
- 이 부분을 여기에 구현하지 않는다면 어떻게 ViewController와 분리해서 구현할 수 있을까 의견이 궁금합니다.

AppleCEO replied:
- 구매 이력이 추가될 때 마지막 구매 이력만 확인해서 이미지 출력하도록 수정하였습니다.
- imageView 라고 풀어서 써봤습니다.
- 고민해보고 말씀드리겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -55,19 +79,71 @@ class ViewController: UIViewController {
+        }
+        if drink == supplyableDrinkList[5] {
+            return "pepsiCoke.jpg"
+        }
```
이렇게 비교문과 이미지 이름을 하드 코딩하는 건 나쁜 표현입니다. 
논리적이고 구조적으로 매칭할 수 있도록 데이터 구조를 만들어보세요.

AppleCEO replied:
음료 메뉴를 저장한 enum 타입에 메소드로 구현해봤습니다.

AppleCEO committed:
추가 - 모서리가 둥근 뷰 생성


음료수 이미지 뷰들이 둥근 뷰로 구성되도록 수정

AppleCEO committed:
추가 - 음료수와 순서를 매칭하는 enum 생성

AppleCEO committed:
수정 - 재고 출력하는 메서드에서 재고가 있는 숫자만 출력하도록 수정

AppleCEO committed:
제거 - get 으로 상태값을 바로 리턴하던 메서드들 제거

AppleCEO committed:
수정 - 판매 목록 또한 하나씩 출력하도록 수정

AppleCEO committed:
제거 - 음료 종류 enum 타입을 생성하면서 안쓰게 된 구조체 제거

AppleCEO committed:
추가 - 구매시 한 음료수 이미지만 추가하는 메서드 구현

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -106,44 +102,35 @@ class ViewController: UIViewController {
+                let imageView = UIImageView(image: image!)
+                imageView.frame = CGRect(x: xCoordinate, y: 575, width: 100, height: 100)
+                self.view.addSubview(imageView)
+        })
```
마지막 하나, SellListPrintable 와 SellDrinkPrintable 프로토콜은 어떻게 다른 역할을 나눠서 하나요?
내부에 이미지 추가하는 코드는 중복된 것 같은데요? 뭐가 다른지 잘 모르겠습니다. 
한꺼번에 여러 개 추가하는 것과 하나 추가하는 것으로 나뉜건가요? 
이 부분만 확인하고 마무리합니다.

kiljh-me replied:
네 조금 저도 정리가 필요하긴 한데요.

AppleCEO replied:
네 처음에 시작할 때 SellListPrintable 를 통해서 전체 판매 목록을 출력해줍니다.
그리고 구매가 하나씩 될 때마다 SellDrinkPrintable 을 사용해서 이미지 하나씩만 추가합니다.

godrm replied:
그러면 여러 개 보내는 걸로 일반화할 수 있지 않나요? 처리하는 코드도 거의 동일해보이는데...

AppleCEO replied:
아.. 네 개선하겠습니다!

AppleCEO replied:
판매 목록 여러 개 출력하는 것을 한개 출력하는 것으로 활용해서 처리해봤습니다.

AppleCEO committed:
수정 - 판매 목록을 출력할 때 하나의 판매 음료수 출력하는 메소드 활용

기존에 두개의 프로토콜로 나뉘어있던 것을 하나로 합쳤습니다.
addSubViewToSellDrinkList 는 addSubView 메서드를 활용해 전체 판매 목록을 출력합니다.
